skimage/morphology/_rolling_ball_cy.pyx

```diff
@@ -0,0 +1,234 @@
+        The index in the resulting array
+    """
+
+    cdef Py_ssize_t ndim = from_shape.shape[0]
+    cdef Py_ssize_t idx, modulo, pos
+
+    cdef Py_ssize_t multiple_kernel = 1
+    cdef Py_ssize_t multiple_img = 1
+    cdef Py_ssize_t to_index = offset
+
+    for idx in range(ndim-1, -1, -1):
+        modulo = from_index % (from_shape[idx] * multiple_kernel)
+        pos = modulo // multiple_kernel
+        multiple_kernel *= from_shape[idx]
+
+        to_index += pos * multiple_img
+        multiple_img *= to_shape[idx]
+
+    return to_index
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def apply_kernel_nan(DTYPE_FLOAT[::1] img,
+                     DTYPE_FLOAT[::1] kernel,
+                     DTYPE_FLOAT[::1] ellipsoid_intensity,
+                     Py_ssize_t[::1] img_shape,
+                     Py_ssize_t[::1] padded_img_shape,
+                     Py_ssize_t[::1] kernel_shape,
+                     num_threads=None):
+    """
+    apply_kernel_nan(img, kernel, ellipsoid_intensity, img_shape,
+                     padded_img_shape, kernel_shape, num_threads)
+
+    Apply a ND kernel to an ND image.
+
+    This function is the critical piece of code for 
+    `morphology.rolling_ellipsoid`. It was moved to cython for speed.
+
+    Parameters
+    ----------
+    img : (I) ndarray
+        A flat view into a padded image, e.g., from ``numpy.ravel()``.
+    kernel : (K) ndarray
+        A flat view into the kernel, e.g., from ``numpy.ravel()``. Indicates
+        if a pixel lies inside the ellipsoid. `kernel[x,y] == 1` if the pixel
+        is inside, ``kernel[x,y] == np.Inf`` otherwise.
+    ellipsoid_intensity : (K) ndarray
+        A flat view into the intensity of the ellipsoid, e.g., from 
+        ``numpy.ravel()``. Indicates the height/intensity of the ellipsoid at
+        position ``(x,y)``.
+    img_shape : (N) ndarray
+        The shape of the unflattened, unpadded image
+    padded_img_shape : (N) ndarray
+        The shape of the unflattened, padded image
+    kernel_shape : (N) ndarray
+        The shape of the unflattened kernel
+
+    Returns
+    -------
+    out_data : ndarray
+        the estimated background intensity. ``out_data.shape = img_shape``.
+
+    See Also
+    --------
+    rolling_ellipsoid
+    """
+    
+    cdef DTYPE_FLOAT[::1] out_data = np.zeros(np.prod(img_shape.base),
+                                              dtype=img.base.dtype)
+    cdef Py_ssize_t offset, offset_idx, out_data_size, ker_idx, img_idx
+    cdef DTYPE_FLOAT min_value, tmp
+    out_data_size = out_data.size
+    cdef int max_threads
+
+    cdef Py_ssize_t ndim = kernel_shape.shape[0]
+    cdef Py_ssize_t kernel_outer = np.prod(kernel_shape[0:(ndim - 1)])
+    cdef Py_ssize_t kernel_inner = kernel_shape[ndim - 1]
+    cdef Py_ssize_t ker_idx_outer, ker_idx_inner
+
+    if num_threads is None:
+        max_threads = openmp.omp_get_max_threads()
+    else:
+        max_threads = <int> num_threads
```

Comment:
The problem with `multiprocessing.cpu_count()` is that we need to account for OpenMP's `OMP_NUM_THREADS` if `max_threads=None`. Instead of explicitly checking, I let OpenMP sort that out. 

If you happen to know how to pass `max_threads=None` into a `nogil` block, we can remove this and not do any checking :)

Reply:
I see in cython [doc](https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html) that `num_threads` default value in `prange` is `None`, so I think that passing `max_thread=None` should work.
Otherwise, you may do something like:
```python
if num_threads is not None:
    openmp.omp_set_num_threads(num_threads)
```
to avoid specifying `num_threads` in `prange` call.

Reply:
> so I think that passing `max_thread=None` should work.

Except that it doesn't. 💯 If I try to pass in `None` cython (rightfully) complains at compilation that I shouldn't access python variables in a `nogil` block.

The problem with using `openmp.omp_set_num_threads` is that it will modify the number of threads process-wide, which may lead to unexpected behavior if a function further downstream uses OpenMP. Ofc, we could reset the value after the parallel block, but then we don't gain anything over having the `if ... else ...` block.

Reply:
> Except that it doesn't

:sob: 

> The problem with using openmp.omp_set_num_threads is that it will modify the number of threads process-wide, which may lead to unexpected behavior if a function further downstream uses OpenMP.

This is also true for the actual implementation that supposes that the user want to use all the available threads when `num_threads` is `None`!

Let's consider that `num_threads` argument is for advanced users and that we modify `OMP_NUM_THREADS` only if it is set. What do you think?

Reply:
> This is also true for the actual implementation that supposes that the user want to use all the available threads when num_threads is None!

Indeed. (`openmp.omp_get_max_threads()` is not equal to the number of cores, so that one works in this context)

> Let's consider that `num_threads` argument is for advanced users and that we modify `OMP_NUM_THREADS` only if it is set. What do you think?

We don't ever modify `OMP_NUM_THREADS`, we simply account for it if specified. I am not 100% certain, but I think the intended order of preference is/should be something like:

```
> num_threads (rolling_ellipsoid if explicitly specified) 
  > openmp.omp_set_num_threads() (prior in the same process) 
    > OMP_NUM_THREADS (if explicitly specified)
      > omp default (typically equal to cpu_count(), but not always), 
```

We can apparently achieve this effect via `num_threads=0` as default (see: https://github.com/scikit-image/scikit-image/pull/4851#discussion_r477223452).